httpbp/headers.go

```diff
@@ -0,0 +1,188 @@
+package httpbp
+
+import (
+	"context"
+	"net/http"
+)
+
+const (
+	edgeContextHTTPHeaderKey = "X-Edge-Request"
+	parentIDHTTPHeaderKey    = "X-Parent"
+	spanIDHTTPHeaderKey      = "X-Span"
+	spanFlagsHTTPHeaderKey   = "X-Flags"
+	spanSampledHTTPHeaderKey = "X-Sampled"
+	traceIDHTTPHeaderKey     = "X-Trace"
+)
+
+type headerContextKey int
+
+const (
+	edgeContextContextKey headerContextKey = iota
+	traceIDContextKey
+	parentIDContextKey
+	spanIDContextKey
+	spanFlagsContextKey
+	spanSampledContextKey
+)
+
+// HTTPHeaderKey is an Enum used to get HTTP headers from a context object.
+//
+// It is not used as the actual key in the context, these enums are mapped to
+// un-exported values that serve as the actual context keys.
+type HTTPHeaderKey int
+
+const (
+	// EdgeContextHTTPHeader is the key for the raw edge request context
+	EdgeContextHTTPHeader HTTPHeaderKey = iota
+
+	// TraceIDHTTPHeader is the header for the trace ID passed by the caller
+	TraceIDHTTPHeader
+
+	// ParentIDHeader is the header for the parent ID passed by the caller
+	ParentIDHeader
+
+	// SpanIDHeader is the header for the span ID passed by the caller
+	SpanIDHeader
+
+	// SpanFlagsHeader is the header for the span flags passed by the caller
+	SpanFlagsHeader
+
+	// SpanSampledHeader is the header for the sampled flag passed by the caller
+	SpanSampledHeader
+)
+
+var headerKeyToContextKey = map[HTTPHeaderKey]headerContextKey{
+	EdgeContextHTTPHeader: edgeContextContextKey,
+	TraceIDHTTPHeader:     traceIDContextKey,
+	ParentIDHeader:        parentIDContextKey,
+	SpanIDHeader:          spanIDContextKey,
+	SpanFlagsHeader:       spanFlagsContextKey,
+	SpanSampledHeader:     spanSampledContextKey,
+}
+
+// GetHeader returns the HTTP header stored on the context at key.
+func GetHeader(ctx context.Context, key HTTPHeaderKey) (string, bool) {
+	if contextKey, ok := headerKeyToContextKey[key]; ok {
+		if header, ok := ctx.Value(contextKey).(string); ok {
+			return header, true
+		}
+	}
+	return "", false
+}
+
+// Headers is an interface to collect all of the HTTP headers for a particular
+// baseplate resource (spans and edge contexts) into a struct that provides an
+// easy way to convert them into HTTP headers.
+//
+// This interface exists so we can avoid having to do runtime checks on maps to
+// ensure that they have the right keys set when we are trying to sign or verify
+// a set of HTTP headers.
+type Headers interface {
+	// AsHTTPHeaders returns the Headers struct as HTTP headers, a map of header
+	// keys to header values.
+	AsHTTPHeaders() map[string]string
```

Comment:
`http.Header` is actually `map[string][]string`, because you could have more than one of the same header: https://godoc.org/net/http#Header

I'm not sure if we should map it into that type, or call it `AsMap` instead to avoid confusion.